s3stream/src/main/java/com/automq/stream/s3/S3Storage.java

```diff
@@ -616,36 +617,35 @@ private void handleAppendRequest(WalWriteRequest request) {
     }
 
     private void handleAppendCallback(WalWriteRequest request) {
-        suppress(() -> handleAppendCallback0(request), LOGGER);
+        suppress(() -> {
+            final long startTime = System.nanoTime();
+            handleAppendCallback0(request);
+            StorageOperationStats.getInstance().appendCallbackStats.record(TimerUtil.durationElapsedAs(startTime, TimeUnit.NANOSECONDS));
+        }, LOGGER);
     }
 
     private void handleAppendCallback0(WalWriteRequest request) {
-        final long startTime = System.nanoTime();
-        List<WalWriteRequest> waitingAckRequests;
-        Lock lock = getStreamCallbackLock(request.record.getStreamId());
-        lock.lock();
-        try {
-            waitingAckRequests = callbackSequencer.after(request);
-            waitingAckRequests.forEach(r -> r.record.retain());
-            for (WalWriteRequest waitingAckRequest : waitingAckRequests) {
-                boolean full = deltaWALCache.put(waitingAckRequest.record);
-                waitingAckRequest.confirmed = true;
-                if (full) {
-                    // cache block is full, trigger WAL upload.
-                    uploadDeltaWAL();
-                }
+
+        List<WalWriteRequest> waitingAckRequests = callbackSequencer.after(request);
+        if (waitingAckRequests.isEmpty()) {
+            return;
+        }
+        for (WalWriteRequest waitingAckRequest : waitingAckRequests) {
+            waitingAckRequest.record.retain();
+            boolean full = deltaWALCache.put(waitingAckRequest.record);
```

Comment:
Currently, LogCache relies on external guarantees to put individual Stream Records in order, moving the lock to WALCallbackSequencer#after will have thread concurrency safety issues.